AdminWebsite/AdminWebsite/Security/ClaimsMemoryCacheProvider.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace AdminWebsite.Security
+{
+    public interface IClaimsCacheProvider
+    {
+        Task<IEnumerable<Claim>> GetOrAdd(string key, Func<string, Task<IEnumerable<Claim>>> valueFactory);
+    }
+
+    public class MemoryClaimsCacheProvider : IClaimsCacheProvider
+    {
+        private readonly ConcurrentDictionary<string, Task<IEnumerable<Claim>>> _cache;
```

Comment:
Feels jiffy to cache tasks. A task is a function right, which feels like you can't cache that, you can only cache results of something no?

Reply:
No, Task is just an object that defines the work that needs to be done. When you get the task back from the cache, it will be executed, whether awaited or not.  And here we await it.

Reply:
I have changed this to cache the Task<UserRole>, and then parse claims afterwards.